File: app/src/main/java/com/worldventures/dreamtrips/wallet/analytics/oncard/SendOnCardAnalyticsCommand.java
package com.worldventures.dreamtrips.wallet.analytics.oncard;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.service.WalletAnalyticsInteractor;
import com.worldventures.dreamtrips.wallet.analytics.oncard.action.SmartCardAnalyticsAction;
import com.worldventures.dreamtrips.wallet.analytics.oncard.action.SmartCardPaymentAction;
import com.worldventures.dreamtrips.wallet.analytics.oncard.action.SmartCardUserAction;
import com.worldventures.dreamtrips.wallet.domain.storage.disk.RecordsStorage;
import com.worldventures.dreamtrips.wallet.service.WalletSocialInfoProvider;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import io.techery.janet.smartcard.model.analytics.AnalyticsLog;
import rx.Observable;

@CommandAction
public class SendOnCardAnalyticsCommand extends Command<Void> implements InjectableAction {

   @Inject WalletAnalyticsInteractor analyticsInteractor;
   @Inject RecordsStorage recordsStorage;
   @Inject WalletSocialInfoProvider socialInfoProvider;

   private final List<AnalyticsLog> analyticsLogs;

   public SendOnCardAnalyticsCommand(List<AnalyticsLog> analyticsLogs) {
      this.analyticsLogs = analyticsLogs;
   }

   @Override
   protected void run(CommandCallback<Void> callback) throws Throwable {
      if (analyticsLogs.isEmpty()) {
         callback.onSuccess(null);
         return;
      }

      Observable.from(analyticsLogs)
            .map(SmartCardAnalyticsAction::from)
            .filter(analyticsAction -> analyticsAction != null)
            .doOnNext(analyticsAction -> {
               if (analyticsAction instanceof SmartCardPaymentAction) {
                  fillRecordDetails((SmartCardPaymentAction) analyticsAction);
               } else if (analyticsAction instanceof SmartCardUserAction) {
                  fillUserDetails((SmartCardUserAction) analyticsAction);
               }
            })
            .doOnNext(analyticsAction -> analyticsInteractor.analyticsActionPipe().send(analyticsAction))
            .toList()
            .map(analyticsAction -> (Void) null)
            .subscribe(callback::onSuccess, callback::onFail);
   }

   private void fillRecordDetails(SmartCardPaymentAction analyticsAction) {
      int paymentCardId = analyticsAction.getRecordId();
      if (paymentCardId >= 0) {
         final String storedRecordId = String.valueOf(paymentCardId);
         analyticsAction.setRecord(Queryable.from(recordsStorage.readRecords())
               .firstOrDefault(record -> storedRecordId.equals(record.id())));
      }
   }

   private void fillUserDetails(SmartCardUserAction analyticsAction) {
      analyticsAction.setUserId(socialInfoProvider.userId());
   }
}

File: wallet/src/main/java/com/worldventures/wallet/analytics/oncard/SendOnCardAnalyticsCommand.kt
package com.worldventures.wallet.analytics.oncard

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.analytics.oncard.action.SmartCardAnalyticsAction
import com.worldventures.wallet.analytics.oncard.action.SmartCardPaymentAction
import com.worldventures.wallet.analytics.oncard.action.SmartCardUserAction
import com.worldventures.wallet.domain.storage.disk.RecordsStorage
import com.worldventures.wallet.service.WalletAnalyticsInteractor
import com.worldventures.wallet.service.WalletSocialInfoProvider
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import io.techery.janet.smartcard.model.analytics.AnalyticsLog
import rx.Observable
import javax.inject.Inject

@CommandAction
class SendOnCardAnalyticsCommand(private val analyticsLogs: List<AnalyticsLog>) : Command<Void>(), InjectableAction {

   @Inject lateinit var analyticsInteractor: WalletAnalyticsInteractor
   @Inject lateinit var recordsStorage: RecordsStorage
   @Inject lateinit var socialInfoProvider: WalletSocialInfoProvider

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Void>) {
      if (analyticsLogs.isEmpty()) {
         callback.onSuccess(null)
         return
      }

      Observable.from(analyticsLogs)
            .map { SmartCardAnalyticsAction.from(it) }
            .filter { analyticsAction -> analyticsAction != null }
            .doOnNext { analyticsAction ->
               if (analyticsAction is SmartCardPaymentAction) {
                  fillRecordDetails(analyticsAction)
               } else if (analyticsAction is SmartCardUserAction) {
                  fillUserDetails(analyticsAction)
               }
            }
            .doOnNext { analyticsAction -> analyticsInteractor.analyticsActionPipe().send(analyticsAction) }
            .toList()
            .map<Void> { _ -> null }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   private fun fillRecordDetails(analyticsAction: SmartCardPaymentAction) {
      val paymentCardId = analyticsAction.recordId
      if (paymentCardId >= 0) {
         val storedRecordId = paymentCardId.toString()
         analyticsAction.setRecord(recordsStorage.readRecords().find { record -> storedRecordId == record.id })
      }
   }

   private fun fillUserDetails(analyticsAction: SmartCardUserAction) {
      analyticsAction.setUserId(socialInfoProvider.userId()!!)
   }
}
